training/src/main/scala/com/airbnb/aerosolve/training/LowRankLinearTrainer.scala

```diff
@@ -0,0 +1,327 @@
+                    }
+                  }
+                }
+              }
+            }
+          }
+        })
+        gradient.iterator
+      })
+      .reduceByKey((a, b) => GradientUtils.sumGradients(a,b))
+      .collectAsMap
+      .toMap
+  }
+
+  def trainAndSaveToFile(sc : SparkContext,
+                         input : RDD[Example],
+                         config : Config,
+                         key : String) = {
+    val model = train(sc, input, config, key)
+    TrainingUtils.saveModel(model, config, key + ".model_output")
+  }
+
+  private def parseTrainingOptions(config : Config) : LowRankLinearTrainerOptions = {
+    LowRankLinearTrainerOptions(
+      loss = config.getString("loss"),
+      iterations = config.getInt("iterations"),
+      rankKey = config.getString("rank_key"),
+      lambda = config.getDouble("lambda"),
+      subsample = config.getDouble("subsample"),
+      minCount = config.getInt("min_count"),
+      cache = Try(config.getString("cache")).getOrElse(""),
+      solver = Try(config.getString("solver")).getOrElse("rprop"),
+      embeddingDimension = config.getInt("embedding_dimension"),
+      rankLossType = Try(config.getString("rank_loss")).getOrElse("uniform")
+    )
+  }
+
+  private def pickRandomOtherLabel(model: LowRankLinearModel,
+                                   posLabels: java.util.Map[java.lang.String, java.lang.Double],
+                                   posIdx: Int,
+                                   posMargin: Double,
+                                   rnd: Random,
+                                   dim: Int) : (Int, String, Double, Int) = {
+    // Pick a random other label. This can be a negative or a positive with a smaller margin.
+    var negIdx = rnd.nextInt(dim)
+    var negLabel = model.getLabelDictionary.get(negIdx).label
+    var negMargin : Double = if (posLabels.containsKey(negLabel)) posLabels.get(negLabel) else 0.0
+    var iter = 0
+    while ((negIdx == posIdx || negMargin > posMargin) && iter < dim) {
+      // we only want to pick a label that has smaller margin, we try at most dim times
+      // we try this for at most dim times
+      negIdx = rnd.nextInt(dim)
+      negLabel = model.getLabelDictionary.get(negIdx).label
+      negMargin = if (posLabels.containsKey(negLabel)) posLabels.get(negLabel) else 0.0
+      iter += 1
+    }
+    (negIdx, negLabel, negMargin, iter)
+  }
+
+  private def rankToLoss(k : Int, dim: Int, lossType: String): Float = {
+    lossType match {
+      case "uniform" => {
+        uniformRankLoss(k, dim)
+      }
+      case "non_uniform" => {
+        nonUniformRankLoss(k, dim)
+      }
+      case _ => {
+        1.0f
+      }
+    }
+  }
+
+  private def uniformRankLoss(k: Int, dim: Int): Float = {
+    // dim is the number of classes
+    // assume we always have dim > 1
+    k * 1.0f / (dim - 1.0f)
+  }
+
+  private def nonUniformRankLoss(k: Int, dim: Int): Float = {
```

Comment:
this term is defined in the wsabie paper, I guess all three losses are 'WARP' loss.